completion/all.go

```diff
@@ -6,74 +6,81 @@ import (
 	"os"
 	"os/user"
 	"path/filepath"
+	"runtime"
 	"strings"
 	"text/template"
 
 	"github.com/coder/serpent"
 )
 
+type Shell interface {
+	Name() string
+	InstallPath() (string, error)
+	UsesOwnFile() bool
+	WriteCompletion(io.Writer) error
+}
+
 const (
-	BashShell  string = "bash"
-	FishShell  string = "fish"
-	ZShell     string = "zsh"
-	Powershell string = "powershell"
+	ShellBash       string = "bash"
+	ShellFish       string = "fish"
+	ShellZsh        string = "zsh"
+	ShellPowershell string = "powershell"
 )
 
-var shellCompletionByName = map[string]func(io.Writer, string) error{
-	BashShell:  generateCompletion(bashCompletionTemplate),
-	FishShell:  generateCompletion(fishCompletionTemplate),
-	ZShell:     generateCompletion(zshCompletionTemplate),
-	Powershell: generateCompletion(pshCompletionTemplate),
+func ShellByName(shell, programName string) Shell {
+	switch shell {
+	case ShellBash:
+		return Bash(runtime.GOOS, programName)
+	case ShellFish:
+		return Fish(runtime.GOOS, programName)
+	case ShellZsh:
+		return Zsh(runtime.GOOS, programName)
+	case ShellPowershell:
+		return Powershell(runtime.GOOS, programName)
+	default:
+		return nil
+	}
 }
 
 func ShellOptions(choice *string) *serpent.Enum {
-	return serpent.EnumOf(choice, BashShell, FishShell, ZShell, Powershell)
-}
-
-func WriteCompletion(writer io.Writer, shell string, cmdName string) error {
-	fn, ok := shellCompletionByName[shell]
-	if !ok {
-		return fmt.Errorf("unknown shell %q", shell)
-	}
-	fn(writer, cmdName)
-	return nil
+	return serpent.EnumOf(choice, ShellBash, ShellFish, ShellZsh, ShellPowershell)
 }
 
-func DetectUserShell() (string, error) {
+func DetectUserShell(programName string) (Shell, error) {
```

Comment:
This feels a bit awkward, but we need to provide it to use in the template itself, e.g. `function _name`, and potentially in the install path, e.g. `name.fish`.